src/ui/mod.rs

```diff
@@ -1,262 +1,319 @@
-                0 => 0,
-                n => n-1,
+
+    fn print_state(&mut self, game: &Game, player: usize) -> io::Result<()> {
+        const SEPARATOR: &str = "--------------------";
+        let player = &game.players[player];
+
+        write!(
+            self.output,
+            "{separator}\n\
+            Color: {color}\n\
+            Hand: {hand}\n\
+            {separator}\n\
+            Table: {table}\n\
+            {separator}\n",
+            separator = SEPARATOR,
+            color = player.identity,
+            hand = player.hand,
+            table = game.table,
+        )
+    }
+
+    fn prompt_bool(&mut self, message: impl fmt::Display) -> io::Result<bool> {
+        let mut buffer = String::new();
+
+        loop {
+            buffer.clear();
+            write!(self.output, "{} (yes/no) > ", message)?;
+            self.output.flush()?;
+            self.input.read_line(&mut buffer)?;
+
+            break match buffer.trim() {
+                "yes" => Ok(true),
+                "no" => Ok(false),
+                _ => {
+                    write!(self.output, "Please enter 'yes' or 'no' > ")?;
+                    continue;
+                }
             };
-            commons.swap(card_num, index);
-        }
-        else {
-            write!(output, "Not a valid selection! Please try again.\n");
         }
     }
-    commons.split_off(num_commons);
-    commons
-}
 
-fn select_card_hand(output: &mut impl Write, player: &Player) -> Result<usize, String> {
-    loop {
-        write!(output, "{}\n", player.hand);
-        write!(output, "> ");
-        output.flush().unwrap();
-        let card = read_uint_from_user();
-        if card > player.hand.len() {
-            write!(output, "{} is not a valid selection!\n", card);
-            continue;
+    fn prompt_usize(&mut self, message: impl fmt::Display) -> io::Result<usize> {
+        let mut buffer = String::new();
+
+        loop {
+            buffer.clear();
+            write!(self.output, "{}\n> ", message)?;
+            self.output.flush()?;
+            self.input.read_line(&mut buffer)?;
+
+            if let Ok(r) = buffer.trim().parse::<usize>() {
+                break Ok(r);
+            }
+            writeln!(self.output, "What you entered is not an unsigned integer! Please try again.")?;
         }
-        if card == player.hand.len() {
-            break Err(String::from("Exiting hand selection.."));
+    }
+
+    fn prompt_hand_selection(&mut self, game: &Game, player: usize, message: impl fmt::Display) -> io::Result<usize> {
+        let player = &game.players[player];
+
+        loop {
+            writeln!(self.output, "{}", message)?;
+            let selection = self.prompt_usize(&player.hand)?;
+
+            if selection >= player.hand.len() {
+                writeln!(self.output, "{} is not a valid selection!", selection)?;
+                continue;
+            }
+
+            break Ok(selection);
```

Comment:
How will players quit hand selection? I guess they could make an invalid selection, but what if they only have cards to make a valid selection? I suppose they could just select the same card. If so should that behaviour be stated?